bokehjs/src/lib/models/glyphs/image_base.ts

```diff
@@ -67,20 +66,19 @@ export abstract class ImageBaseView extends XYGlyphView {
       if (indices != null && indices.indexOf(i) < 0)
         continue
 
-      let img: Arrayable<number>
-      if (this._image_shape != null && this._image_shape[i].length > 0) {
-        img = this._image[i] as Arrayable<number>
-        const shape = this._image_shape[i]
-        this._height[i] = shape[0]
-        this._width[i] = shape[1]
+      const img = this._image[i]
+      let flat_img: Arrayable<number>
+      if (is_NDArray(img)) {
+        flat_img = img
+        this._height[i] = img.shape[0]
+        this._width[i] = img.shape[1]
       } else {
-        const _image = this._image[i] as number[][]
-        img = concat(_image)
-        this._height[i] = _image.length
-        this._width[i] = _image[0].length
+        flat_img = concat(img)
+        this._height[i] = img.length
+        this._width[i] = img[0].length
```

Comment:
We could deprecate this and remove for 3.0 now that NumPy is a hard dependency there's no reason to not just require real arrays for images

Reply:
I started https://github.com/bokeh/bokeh/issues/10090 for this, as it doesn't seem something trivial to do.